[dynamo] Support weakref objects

anijain2305 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #128533

Fixes https://github.com/pytorch/pytorch/issues/125720

I was earlier worried that DELETE_* or STORE_* on referent values should result in a graph break, because they could invalidate the weak ref. But then @zou3519 pointed out that weakref invalidation will happen EVENTUALLY, CPython provides no guarantees when the weakref will be invalidated (even when the user calls del x and x is the last reference).

So any code that relies on del x to invalidate the weakref of x right away is BAD code. CPython provide no guarantees. Therefore we can (ab)use this nuance, and can just ignore DELETE_* or STORE_* on the referent objects.

The only corner case is when Dynamo is reconstructing the weakref object. Dynamo will have a hard time being correct here, so just SKIP_FRAME on such a case. This is rare.


Cpython notes
1) https://docs.python.org/3/library/weakref.html
2) https://docs.python.org/3/reference/datamodel.html#index-2

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

anijain2305 committed:
[dynamo] Support weakref objects

[ghstack-poisoned]

anijain2305 commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -3247,13 +3292,13 @@ void install_tensor_aliasing_guard(
 void install_no_tensor_aliasing_guard(
     const py::list& guard_managers,
-    py::list tensor_names,
+    const py::list& tensor_names,
```
Linter error - unrelated to the PR

williamwen42 commented:
Add a test where the weakref'd object is deleted? e.g.
```python
x = Obj()
ref = weakref.ref(x)
opt_fn(ref, inp)
del x
opt_fn(ref, inp)
```

williamwen42 reviewed (commented):
Add a test where the weakref'd object is deleted? e.g.
```python
x = Obj()
ref = weakref.ref(x)
opt_fn(ref, inp)
del x
opt_fn(ref, inp)
```

anijain2305 committed:
Update on "[dynamo] Support weakref objects"


Fixes https://github.com/pytorch/pytorch/issues/125720

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

anijain2305 commented:
@williamwen42 added.

williamwen42 commented:
Is there an assumption that a weakref's call (i.e. `ref()`) won't change during tracing or in the middle of running a dynamo-optimized function?

anijain2305 commented:
@williamwen42 Nice catch! Yeah, this test fails

Let me think how to handle it

~~~
import torch
import weakref

def fn(y):
    x = torch.randn(4)
    x_weak = weakref.ref(x)
    if x_weak is not None and x_weak() is not None:
        z =  torch.sin(y)
    else:
        z = torch.cos(y)

    del x
    if x_weak is not None and x_weak() is not None:
        z =  torch.sin(y)
    else:
        z = torch.cos(y)
    return z




y = torch.ones(4)
ref = fn(y)

opt_fn = torch.compile(fn, backend="eager")
res = opt_fn(y)
print(ref, res)
~~~

williamwen42 commented:
I tried thinking about how to support weakrefs a few months ago and ran into the issue of determining when a weakref gets invalidated - it may be the case that weakrefs can be invalidated at anytime - perhaps by a different thread, or by the garbage collector? If we do manage to support detecting when weakrefs are invalidated, we also need to support tracing the finalizer/callback.

anijain2305 commented:
Lol ... this is tricky .. what was I thinking .. converting to draft

mlazos commented:
From chatting offline I think it should be possible to support just getting the value out - that's most of the use cases I've seen.

Tracking the lifetime of the underlying object is much harder and I don't think it's needed.

anijain2305 commented:
@jansel @williamwen42 @mlazos 

I was worried that DELETE_* or STORE_* on referent values should result in a graph break, because they could invalidate the weak ref. But then @zou3519 pointed out that weakref invalidation will happen EVENTUALLY, CPython provides no guarantees when the weakref will be invalidated (even when the user calls `del x` and x is the last reference). 

So any code that relies on `del x` to invalidate the weakref of x right away is BAD code. CPython provide no guarantees. Therefore we can (ab)use this nuance, and can just ignore DELETE_* or STORE_* on the referent objects. 

The only corner case is when Dynamo is reconstructing the `weakref` object. Dynamo will have a hard time being correct here, so just SKIP_FRAME on such a case. This is rare.

anijain2305 committed:
Update on "[dynamo] Support weakref objects"


Fixes https://github.com/pytorch/pytorch/issues/125720

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo] Support weakref objects"


Fixes https://github.com/pytorch/pytorch/issues/125720

I was earlier worried that DELETE_* or STORE_* on referent values should result in a graph break, because they could invalidate the weak ref. But then zou3519 pointed out that weakref invalidation will happen EVENTUALLY, CPython provides no guarantees when the weakref will be invalidated (even when the user calls del x and x is the last reference).

So any code that relies on del x to invalidate the weakref of x right away is BAD code. CPython provide no guarantees. Therefore we can (ab)use this nuance, and can just ignore DELETE_* or STORE_* on the referent objects.

The only corner case is when Dynamo is reconstructing the weakref object. Dynamo will have a hard time being correct here, so just SKIP_FRAME on such a case. This is rare.


Cpython notes
1) https://docs.python.org/3/library/weakref.html
2) https://docs.python.org/3/reference/datamodel.html#index-2

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo] Support weakref objects"


Fixes https://github.com/pytorch/pytorch/issues/125720

I was earlier worried that DELETE_* or STORE_* on referent values should result in a graph break, because they could invalidate the weak ref. But then zou3519 pointed out that weakref invalidation will happen EVENTUALLY, CPython provides no guarantees when the weakref will be invalidated (even when the user calls del x and x is the last reference).

So any code that relies on del x to invalidate the weakref of x right away is BAD code. CPython provide no guarantees. Therefore we can (ab)use this nuance, and can just ignore DELETE_* or STORE_* on the referent objects.

The only corner case is when Dynamo is reconstructing the weakref object. Dynamo will have a hard time being correct here, so just SKIP_FRAME on such a case. This is rare.


Cpython notes
1) https://docs.python.org/3/library/weakref.html
2) https://docs.python.org/3/reference/datamodel.html#index-2

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

anijain2305 commented:
@pytorchbot merge
